### DISCUSSION TITLE
Source Build and Multiple SDK bands - feedback wanted

### DISCUSSION THREAD

**Discussion by premun** [dotnet/source-build#3524] [General]
## Multiple SDK bands and Source Build

As part of the [Unified Build effort](https://github.com/dotnet/arcade/tree/main/Documentation/UnifiedBuild) on whose status we've recently sent [an update](https://github.com/dotnet/source-build/discussions/3518), we've already changed how we ship the .NET sources to our various distro partners. Starting with .NET 8.0 previews, we have switched to using a mono-repo (the [VMR](https://github.com/dotnet/dotnet)) for this purpose and we aim to continue using it for .NET 9.0 and beyond.

Once we make .NET 8.0 generally available, we expect to also use the VMR to develop and ship multiple SDK bands (e.g. 8.0.100 and 8.0.200). This is a change from the current situation where only the 100th (.NET 7.0.1xx) band is source-buildable. We aim to bring the Source Build experience on par this way.

## Feedback wanted

As this change will have impact on how .NET is consumed by the various distro maintainers, we seek feedback from the community. We have described and analyzed the problematics in a design document for which we have opened a [Pull Request](https://github.com/dotnet/arcade/pull/13720). The PR also contains two proposals on how we can implement the solution which are compared.  
This is where you come as we're interested in opinions on the following:

- Are you, as a distro maintainer, OK with the proposed solution?
- Are you interested in consuming multiple SDK bands? Or would one band be enough? Which one would that be?

Please, have a look at the [PR](https://github.com/dotnet/arcade/pull/13720) which explains what the SDK bands are, what we plan to do about them and let us know!

cc @dotnet/distro-maintainers

**Comment by alucryd** [comment]
As far as Arch Linux is concerned, we will most likely stick to one band, ideally the one that is more advanced/feature-packed (ie more popular/attractive to users).

For instance, right now arch users are flagging our current SDK out of date at least once a week, expecting me to push 3xx instead of 1xx. This has become quite annoying over time, so being able to build the 8.0 equivalent of the current 3xx when it comes out would be greatly appreciated.

**Reply by mmitche** [reply]
@alucryd Thanks for the feedback, that's super valuable information. I have another question regarding potential build workflows and whether they work with your infra.

When we (MSFT) build multiple SDKs today, we essentially build the runtime components of .NET, then flow those outputs to the SDK repositories, where each branch in that repo corresponds to a single SDK band. So for instance, if we have release/8.0 for the runtime branch, then we'd have `release/8.0.1xx` for the 100 band SDK, `release/8.0.2xx` for the 200 band SDK, and so on. It ends up being pretty important that the shared runtime components only be built once for all SDK bands.

From a Linux source-build workflow, today the release/8.0.1xx-previewN branches we spin off for previews enable a single-commit build of the shared runtime and the 1xx band. Check out the tag we provide, and the full set of outputs is produced. When we add in additional SDK bands, managing this as a single commit, single build invocation becomes much tougher.

Our proposal for supporting the additional bands in Linux source-build looks to treat the additional SDK bands as being closer to separate products built on top of the 1xx band and shared runtime. It would look something like this, if you wanted to build the 3xx SDK:

1. Build the full 1xx repository like you do today.
2. For each SDK band you want to ship (note that you do not need to ship the 1xx band):
  1. Check out the dotnet/dotnet branch for that band (e.g. release/8.0.2xx)
  2. Build that band, using the build outputs of the 1xx build.

This ensures that:
- The shared runtime components are not rebuilt for each band.
- The infrastructure to build additional bands remains simple
- The build doesn't expand unnecesarily as new bands are added in servicing.

Questions? Comments?

**Reply by omajid** [reply]
How does this interact with a version N requiring the N-1version to build? For example, if we are are only going to build 8.0.201 (and lets assume that combo is building 8.0.111 and 8.0.201), what do we need to have to be able to build things offline? Will we need both ~~8.0.111~~ 8.0.110 and 8.0.200? Or will 8.0.200 be enough by itself to build both?

**Reply by alucryd** [reply]
Building in incremental stages is fine, and pretty cool if we do decide to build more bands in the future since it currently takes about an hour to build the whole SDK on our beastly build farm.

Elaborating on why I would prefer packaging a single band, splitting them in different packages would mean including 1xx, 2xx and so on in the package name, given the dynamic nature of these bands we might have to drop/add packages for different major releases, which goes against the KISS nature of Arch. Another solution would be to bundle them all in one package, but if bands are incremental upgrades of each other it sounds like shipping the highest one would be essentially that.

**Reply by mmitche** [reply]
All good questions. Here are some basic facts:

- Currently, there is no _guarantee_ that you can use a 1xx SDK to build the newer SDK bands. It's likely that newer SDK bands take dependencies on new features that are introduced in those bands. However, barring unexpected bugs, The N  SDK should always build with N-1 of the same SDK band. So 8.0.202 with 8.0.201.
- Currently there is no _guarantee_ that you can use a 2xx+ SDK to build the runtime. I think it's likely that it will work most of the time, and should work with small modifications (e.g. disable a warning or analyzer), but we don't guarantee it today. The runtime will always build with N-1 of the 1xx band though. We always want to be building it with a supported toolset, and the 1xx band is supported as long as the runtime is in support.

So that leaves a couple of options for offline builds:
- We treat the runtime and SDK as separate products built from separate repos. Runtime in one repo, SDKs in another. You build the runtime with your 1xx sdk band (n-1 or whatever version you choose), and you build the 2xx+ bands with the 2xx sdk band with the 2xx bits from the previous month (or whatever version you choose). IMO the upshot of this is that it's a bit more natural to choose which SDK bands you want to build on a monthly basis. The downside is that if you only care about 1xx, you require more build invocations than you have today. You need to build multiple source branches to get the outputs you want.
- We treat 1xx as special, since it's what is used to build the runtime on a monthly basis, and the 2xx+ bands as SDK-only, requiring the input from the 1xx build. It means you build a bit extra than what you might want to, but for the 1xx band, it's simpler.

Let me make up some diagrams that can make this a bit clearer and you all can tell me what you think.

**Reply by tmds** [reply]
I wrote down some expectations here: https://github.com/dotnet/arcade/pull/13720/files#r1238167475.

**Reply by tmds** [reply]
> Currently there is no guarantee that you can use a 2xx+ SDK to build the runtime.

This wasn't spelled out before. It means it's required to build a 1xx SDK, even if the interest of the maintainer would be to ship the latest band only.

**Reply by jkotas** [reply]
Right, we are not keen on taking unnecessary churn in runtime servicing to make the runtime build with newer SDKs.

**Reply by mmitche** [reply]
@tmds What are your expectations around the version of the SDK used to build the shipping runtime? Must it always be up to date (for security or other policies?). Barring bugs, you could theoretically use 8.0.100 to build runtimes forever.

**Reply by tmds** [reply]
A maintainer will want to use the previous build of the SDK band to build the next.

**Reply by mmitche** [reply]
So that tells me that keeping the 1xx band alongside the runtime, then the 2xx+ bands in separate branches, is the right way to go rather than separating runtime and sdk entirely. This is because you would always need to build the latest 1xx to build the next runtime. While you could probably continue to use an older sdk, it's undesirable, so separating the repos doesn't reduce monthly workload for those only concerned about 2xx.

**Reply by mmitche** [reply]
@alucryd @tmds @omajid This is what I think the build flows would look like:

Initial 1xx SDK (no change from today)

```mermaid
flowchart LR
    bootstrap1xxSDK[MSFT built 8.0.100 SDK]
    initial1xxSdkBuild{Build 8.0.100 SDK}
    initial1xxSdk[8.0.100 SDK from MSFT SDK]

    offline1xxSdkBuild{Build 8.0.100 SDK}
    offline1xxSdk[8.0.100 SDK from source-built SDK]

    bootstrap1xxSDK-->initial1xxSdkBuild-->initial1xxSdk
    initial1xxSdk-->offline1xxSdkBuild-->offline1xxSdk
```

Month when a 200 band goes RTM.

```mermaid
flowchart LR
    bootstrap2xxSDK[MSFT-built 8.0.200 SDK]
    previouslySB1xxSDK["Previously source-built 8.0.10(N-1) SDK"]
    offline1xxSdkBuild{Build 8.0.10N}
    offline1xxSdk[8.0.10N SDK from source-built SDK]
    initial2xxSdkBuild{Build 8.0.200 SDK}
    initial2xxSdk[8.0.200 SDK from MSFT SDK]
    offline2xxSdkBuild{Build 8.0.200 SDK}
    offline2xxSdk[8.0.200 SDK from source-built SDK]

    
    previouslySB1xxSDK-->offline1xxSdkBuild-->offline1xxSdk
    offline1xxSdk-->initial2xxSdkBuild
    bootstrap2xxSDK-->initial2xxSdkBuild-->initial2xxSdk
    initial2xxSdk-->offline2xxSdkBuild-->offline2xxSdk
    offline1xxSdk-->offline2xxSdkBuild
```

- The runtime outputs from the 1xx SDK build feed into the 2xx SDK build

Monthly updates to 1xx and 2xx

```mermaid
flowchart LR
    previouslySB1xxSDK["Previously source-built 8.0.10(N-1) SDK"]
    previouslySB2xxSDK["Previously source-built 8.0.20(N-1) SDK"]
    offline1xxSdkBuild{Build 8.0.10N}
    offline1xxSdk[8.0.10N SDK from source-built SDK]
    offline2xxSdkBuild{Build 8.0.200 SDK}
    offline2xxSdk[8.0.200 SDK from source-built SDK]

    
    previouslySB1xxSDK-->offline1xxSdkBuild-->offline1xxSdk
    previouslySB2xxSDK-->offline2xxSdkBuild-->offline2xxSdk
    offline1xxSdk-->offline2xxSdkBuild
```

- The runtime outputs from the 1xx SDK build feed into the 2xx SDK build

**Reply by mmitche** [reply]
A couple more possible flows. It's possible that the 1xx SDK will build the 2xx+ SDKs. If the newer SDKs do not take dependencies on their new features, then bootstrapping 200 and monthly updates look the same:

```mermaid
flowchart LR
    previouslySB1xxSDK["Previously source-built 8.0.10(N-1) SDK"]
    offline1xxSdkBuild{Build 8.0.10N}
    offline1xxSdk[8.0.10N SDK from source-built SDK]
    offline2xxSdkBuild{Build 8.0.200 SDK}
    offline2xxSdk[8.0.200 SDK from source-built SDK]
    
    previouslySB1xxSDK-->offline1xxSdkBuild-->offline1xxSdk
    previouslySB1xxSDK-->offline2xxSdkBuild-->offline2xxSdk
    offline1xxSdk-->offline2xxSdkBuild
```

**Comment by omajid** [comment]
Is there any description of the release/support timelines for each feature band? I understand that 8.0.1xx will be supported for the entire lifetime of 8.0. Is that correct? What about other feature bands? When do they release and how long will they be supported for?

**Reply by jkotas** [reply]
This is explained in https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Managing-SDK-Bands.md#band-lifecycle

**Reply by omajid** [reply]
Thank you. I am still unclear on the actual dates that a 8.0.2xx band would be GA'ed and EOL'ed. Are those dates known in advance and public? I also couldn't find a rough date on how long a theoretically 8.0.2xx and 8.0.3xx branch might be supported for (6 months, 12 months, 18 months, etc). This is clearer for 8.0.1xx ("All of .NET 8's lifecycle") but not for these newer SDK feature bands.

**Reply by jkotas** [reply]
8.0.2xx+ bands are tied to Visual Studio releases. The releases dates are not known in advance and public. The lifecycle of these bands is tied to Visual Studio releases. They are supported as long as the Visual Studio versions that they are attached to are supported. For example, https://github.com/dotnet/installer/tree/release/6.0.2xx is no longer maintained for this reason.

**Reply by omajid** [reply]
Thanks. Am I reading https://learn.microsoft.com/en-us/visualstudio/productinfo/vs-servicing correctly which suggests that VS release are supported for around 18 months, and so these feature bands would also be supported for around that long?

**Reply by jkotas** [reply]
> suggests that VS release are supported for around 18 months

That is for LTSC channel. For example, the lifetime 6.0.2xx that was attached to current channel only was much shorter.

**Reply by marcpopMSFT** [reply]
This table will help you project out what the lifecycle of the 8.0 feature bands of the SDK will look like: https://learn.microsoft.com/en-us/dotnet/core/porting/versioning-sdk-msbuild-vs#lifecycle

As Jan noted, the feature bands align with Visual Studio versions. While I don't know when 18.0 will ship, I think it's safe to assume that Visual Studio will continue to have quarterly releases which means .NET SDKs will have new feature bands each quarter going from 1xx to 4xx.

We support each SDK band for the life of the VS version is ships in (so 18 months for LTSC and 3 months otherwise). We treat 1xx and 4xx special though with 1xx supported for source build customers for the life of the runtime (ie we won't continue to ship the Windows or Microsoft versions once we are past the Visual Studio support cycle for a 1xx version) and 4xx for the life of the matching .NET runtime for all customers (so the highest number feature band is always in support for that runtime). Does that make sense then?

**Comment by ayakael** [comment]
From Alpine Linux's perspective, we'd probably keep our `edge` repository up to the latest feature band, but maintain the same feature bands after releases.

That said, I do share omajid's questions as to N+1. I also wonder about the possibility of decoupling runtime package from the rest of source-build. Since all feature bands share the same runtime, it seems desirable to only build runtime once.

**Reply by mmitche** [reply]
The main issue there is that you need the 1xx SDK to build the runtime. If you want to keep up to date and use the latest 1xx SDKs to build the runtime every month, you would need to build the 1xx SDK every month anyway. Note that from a build time standpoint, the runtime components take most of the time.

The runtime is still built only once, just with the 1xx band. Then those outputs would be fed as previously source-built artifacts to the 2xx+ build(s).

**Reply by ayakael** [reply]
Circling back to this, it's unfortunate that runtime will need 1xx SDK to build. I would want to ditch 1xx SDK the second 2xx came out. But alas, supporting 2xx, 3xx, and more also means to keep on supporting 1xx. I find that problematic, and would rather have the option to stay on the latest feature band to build both SDK and runtime.

**Reply by tmds** [reply]
I had also hoped that would be possible.

The limiting factor is the runtime is only maintained to build with the 1xx band.

Even if upstream doesn't officially support it, I assume it may work in practice.

**Reply by mmitche** [reply]
> Even if upstream doesn't officially support it, I assume it may work in practice.

I think it is **likely** that building the runtime with the 2xx+ SDKs will be functional. Most issues that crop up would be tooling related (new warnings, analyzers).

**Comment by tmds** [comment]
I've added feedback to the PR, mainly: https://github.com/dotnet/arcade/pull/13720/files#r1238128063.

**Comment by omajid** [comment]
With my Fedora maintainer hat on, I would be interested in building and releasing the latest feature band for Fedora. The stability provided by the oldest feature band (eg, 8.0.1xx) wouldn't necessarily be beneficial to Fedora, specially if it the older feature bands are not frequently used by users and are harder-to-install.

If we are only shipping the latest feature band, the always-required 8.0.1xx build would be a bit wasteful - we would always build 8.0.1xx and then throw it away.

If we were to ship multiple SDK versions, I am not sure how upgrade paths should work. In particular, how should the EOL of specific feature bands should be handled. For example, if Fedora was shipping 8.0.1xx, 8.0.2xx, and 8.0.3xx, and we shipped them as separate packages with side-by-side installation, and a user had only 8.0.2xx installed, how should we handle 8.0.2xx EOL? Some options are to leave users running the EOL versions and force-moving users to 8.0.3xx. None of those are both "good" (form a UX point of view) and easy (from a packaging point of view).

**Reply by mmitche** [reply]
@richlander

**Comment by mirespace** [comment]
Hi all,

happy to see all in the same boat, with similar concerns.

This is an excellent opportunity to satisfy Ubuntu's users on their demands about the latest feature bands ("hey, why are you don't shipping x.x.4xx?"). I have questions for different scenarios, also different within the user point of view or the maintainer/packaging one. Sorry in advance if some questions are naive/newbies-kind.


The different options we see are:

   - If we want to give the users the latest features but ship only an SDK band: it seems similar to the actual set of packages (source and binaries) and update paths. This forces the user to go up through .1XX, .2XX, .3XX and so on.
   - If we want to allow the users to choose what band to use: Needs to redefine the packages, the dependencies between them, the default SDK to be installed, the scenarios or install/uninstall another band, updating paths, etc … This needs more time to be studied, with all the particularities already set about cadences, EOLs, source origin/how to get the source code, building needs and process/flow, time expectations of publishing.
   - Stick it simple: only the 1XX band: like now (but maybe not fair for the users).

And, if packaging multiple bands for Ubuntu, then, depending on the release cadence:

   - If all the SDK bands are released on the same day (patch Tuesday), all is OK.
   - If not, the updated SDK bands after patch Tuesday has to wait for the "stable" SDK band (.1xx) to be released again in Ubuntu (next patch Tuesday) due to our restrictions on Stable Releases Updates (SRU process).
   - The SDK bands different from .1XX need to come from a different source package than dotnet6/7/8, to allow the SRU process happens for being released on GA time.Maybe, in this case, the SDK branches scenario on the VMR is better than the side-by-side then, as in the side-by-side we should introduce manual steps once we get the code from the VMR, returning to an old scenario (before the VMR) of recreating a tarball(in this case, for new SDK bands).
   - An SDK band with short EOL could not fit in an Ubuntu LTS series (i.e., Jammy).

From the maintainer's point of view, my questions are:

- Do the different SDK bands work with different versions of the shared components? I mean, I read at "Current code flow" section at Documentation/UnifiedBuild/VMR-Managing-SDK-Bands.md that the bands share components (runtime/aspnetcore/arcade ...), but that it takes time for the components/band to become coherent.. so, Does the GA of the new SDK bands happen when that coherent happens too or could be the case that different SDK bands use different versions of the shared components? And  @mmitche said "It ends up being pretty important that the shared runtime components only be built once for all SDK bands."... so I'm a bit confused about it. I think to recall that the runtime (dotnet-host) is always backwards compatible but that this doesn't happen for the other shared components.

- Taking a look at the graphic on the example of SDK bands -> Shared vs band-specifc components (line 38 in the corresponding .md file), the releases of the SDKs are consecutive if IUIC, in the sense that
first is Y.0.1XX and later will come Y.0.2XX and from this one Y.0.3XX (not necessarily immediately in preview when the immediate lower SDK band is released) and so on. Therefore, we as maintainers will publish first the 1XX band yes or yes when .NET(8/9/...) would be GA, even if we chose to serve upper bands. 

  -> On the building side of this, apart from the offline build flow that constantly goes through build the SDK band .1XX that can be wasted or not (and building once the shared component if they are coherent -so maybe separated packages for this if a .2XX SDK can be built with a .2XX-1 previous version ?-): I read on some of the comments that for build a .2XX band the output of the .1XX is needed (like "bootstrapping" .2XX band with .1XX in the first place? ). Then, to have different bands in different packages for the bands ... do we also need to produce and consume different sdk-source-built-artifacts per band for offline buildings?



For deciding on this, we need to take into consideration the user/developer perspective also, so I have this set of questions:

- Does it makes sense to have several SDK bands installed for the same .NET version?
- Related to the first question, from the graphic on the example of SDK bands -> Shared vs band-specifc components (line 38 in Documentation/UnifiedBuild/VMR-Managing-SDK-Bands.md), it seems Y.0.2XX start at some point from Y.0.1XX (when the correspondent VS minor releases happen and/or the bands become coherent), but then ... do they diverge or Y.0.2XX will be the contents of Y.0.1XX plus the features that come from the VS channel 1Y.2?
- Also, if the bands share "basic features" as exposed above: in a final .NET app, it could have advantages in weight or performance the use of one SDK or another, i.e., if using only features on the .1XX band in my app's code, does build it against .4XX could make my application poorer?

Once those are clarified, deciding what to do will be easier (I hope so!). Thanks!

**Reply by mmitche** [reply]
Thanks for the reply @mirespace. Let me see if I can cover some of this:

> Do the different SDK bands work with different versions of the shared components? 

Servicing releases of SDK bands have aligned releases, and the shared runtime components will be identical between them. Let me see if I can explain how it works a bit more in depth. During the development of a major release, the 100 band SDK is also in development and GAs at the same time the runtime does. After GA, we branch off the SDK repos (dotnet/sdk, dotnet/installer, and dotnet/templating) into a 2xx branch where previews are developed and released from. This branch has its shared runtime components set to the previously released runtime, and the input tooling components (e.g. roslyn and nuget) start flowing in from branches that correspond to the upcoming VS release. This new 2xx branch takes tooling changes until the VS point release is ready to GA, at which point it GA's at the same time. Generally this occurs on the normal Tuesday update schedule, so we start flowing the shared runtime components into both the 2xx and 1xx branches in that lead up to the Tuesday update, with the requirement that both 2xx and 1xx ship the exact same set of shared runtime components. This ensures that a user who installs a 2xx SDK sees the same runtime behavior and level of security that a user that installs 1xx, **and** that a user that installs both 2xx and 1xx side by side doesn't get odd install behavior. Because the runtime installs by default to a shared location, if 2xx and 1xx were to both install, say "8.0.4", but these 8.0.4's were actually not the same product with the same fixes, things would be very confusing as it may be unclear what files actually got laid down on disk.

Once on a monthly servicing cadence, all SDK bands currently in servicing release at the same time, dependent on the same runtime.

> do we also need to produce and consume different sdk-source-built-artifacts per band for offline buildings?

Yes, or at least it's somewhat likely. The key bit to remember is that the runtime is repackaged in the SDK layout, and all SDK bands released together should have the same shared runtime otherwise in theory a customer could get different behavior deployment an app with a 1xx band SDK vs a 2xx SDK. So, since we need to build the runtime once for all bands and 1xx is used to build that runtime, then we'll need a set of previously source-built components corresponding for building 1xx and the shared runtime. Now, for newer bands you will need a different set of previously source-built components **if the newer band requires itself to build**. If 2xx requires 2xx to build, then of course you'd need a set of 2xx outputs to build offline. Currently, it is not guaranteed that 2xx requires 2xx, or 3xx requires 3xx. Often only 1xx is required. It is mainly about whether the SDK tooling components (nuget, roslyn, etc.) take dependencies on features that only show up in 2xx during the preview development period. If they decide to do so, then they'll update their SDK to use a new version. Otherwise, the SDK will remain at 1xx. If the SDK required to build 2xx's components stayed at 1xx, then I think you could theoretically use the same set of input previously source-built artifacts.

TLDR:
- The 1xx band + runtime will require the previously source-built artifacts of a 1xx build to build offline.
- The 2xx, 3xx, etc.+ bands will require the previously source-built artifacts of a 1xx build (because it contains the runtime they are to repackage and depend on)
- I think it is likely, that 2xx, 3xx, etc+ bands will **also** require the previously source-built outputs of their corresponding band to build.

> Does it makes sense to have several SDK bands installed for the same .NET version?

This is not an uncommon scenario. Since global.json can target different versions of the SDK, SxS installations do happen.

> do they diverge or Y.0.2XX will be the contents of Y.0.1XX plus the features that come from the VS channel 1Y.2?

We take care to ensure that the functionality of 2xx always includes 1xx. 2xx would include newer tooling, but fixes made to older versions of tooling, and the SDK, that make it into 1xx, should make it into 2xx.

> Also, if the bands share "basic features" as exposed above: in a final .NET app, it could have advantages in weight or performance the use of one SDK or another, i.e., if using only features on the .1XX band in my app's code, does build it against .4XX could make my application poorer?

Since the runtime is the same between all of these bands, the SDK has limited power to alter performance in newer bands. Generally, the kinds of features that show up in newer bands are centered around development tooling and not runtime behavior.

@jkotas @marcpopMSFT @jaredpar may have some comments on that least bit.

**Reply by jaredpar** [reply]
> ... if using only features on the .1XX band in my app's code, does build it against .4XX could make my application poorer?

No. The feature bands are centered around developer tooling and supporting the design time experience in our editors (VS and VS Code). Moving to a higher feature band should not negatively impact the performance of your application. 

Or said another way, if it did negatively impact the performance of your application it's something we would want to take a look at because most likely that would represent a bug.
